### ISSUE TITLE
[feature request] Add boolean reduce operations

### ISSUE THREAD

**Issue by zasdfgbnm** [pytorch/pytorch#7539]
Hi, I just want to ask if it is contribute welcome for adding two boolean reduce operations `torch.all` and `torch.any`. The API is the the same as `torch.sum` and `torch.prod`. These two ops treat the input tensor as boolean tensor and compute the `and` and `or` along the the specified dimension.

`torch.all` can be implemented as `torch.prod`, but implementing `torch.any` as `torch.sum` is dangerous because it is possible that there happen to be 256 1s in the tensor, which would gives a 0 instead of some non-zero value and in my application, I'm using a workaround for `torch.any` that looks like
```python
torch.sum(x.dtype(torch.float), dim) > 0
```

**Comment by apaszke** [comment]
We already have them, although they are present only as methods of Boolean tensors (`.all()`, `.any()`). Since they’re limited to a single type only I think it’s better to avoid adding them to the main namespace. Thanks for the proposal!

**Comment by zasdfgbnm** [comment]
Ahh, version 0.5 add dim support. Thanks!
